GitUI/UserControls/RevisionGrid/Columns/CommitIdColumnProvider.cs

```diff
@@ -33,37 +34,59 @@ public override void ApplySettings()
             Column.Visible = AppSettings.ShowObjectIdColumn;
         }
 
-        public override void OnCellPainting(DataGridViewCellPaintingEventArgs e, GitRevision revision, int rowHeight, in CellStyle style)
+        private int GetCharLengthForColumnWidth(int width)
         {
-            Font monospaceFont = style.MonospaceFont;
+            Font monospaceFont = AppSettings.MonospaceFont;
             if (!_widthByLengthByFont.TryGetValue(monospaceFont, out int[] widthByLength))
             {
                 widthByLength = Enumerable.Range(0, ObjectId.Sha1CharCount + 1).Select(c => TextRenderer.MeasureText(new string('8', c), monospaceFont).Width).ToArray();
 
                 _widthByLengthByFont[monospaceFont] = widthByLength;
             }
 
-            if (!revision.IsArtificial)
+            int i = Array.FindIndex(widthByLength, w => w > width);
+
+            if (i == -1 && width >= widthByLength[^1])
             {
-                int i = Array.FindIndex(widthByLength, w => w > Column.Width);
+                return ObjectId.Sha1CharCount;
+            }
+            else if (i > 1)
+            {
+                return i - 1;
+            }
 
-                if (i == -1 && Column.Width > widthByLength[^1])
-                {
-                    _grid.DrawColumnText(e, revision.ObjectId.ToString(), monospaceFont, style.ForeColor, e.CellBounds, useEllipsis: false);
-                }
-                else if (i > 1)
-                {
-                    _grid.DrawColumnText(e, revision.ObjectId.ToShortString(i - 1), monospaceFont, style.ForeColor, e.CellBounds, useEllipsis: false);
-                }
+            return 0;
+        }
+
+        public override void OnCellPainting(DataGridViewCellPaintingEventArgs e, GitRevision revision, int rowHeight, in CellStyle style)
+        {
+            if (e.FormattedValue == null)
+            {
+                return;
             }
+
+            _grid.DrawColumnText(e, (string)e.FormattedValue, style.MonospaceFont, style.ForeColor, e.CellBounds, useEllipsis: false);
+        }
+
+        public override void OnColumnWidthChanged(DataGridViewColumnEventArgs e)
+        {
+            _charCount = GetCharLengthForColumnWidth(e.Column.Width);
         }
 
         public override void OnCellFormatting(DataGridViewCellFormattingEventArgs e, GitRevision revision)
         {
             // Set the grid cell's accessibility text
             if (!revision.IsArtificial)
             {
-                e.Value = revision.ObjectId.ToShortString();
+                if (!_charCount.HasValue)
+                {
+                    _charCount = GetCharLengthForColumnWidth(Column.Width);
+                }
+
+                if (_charCount != 0)
```

Comment:
Should not matter, but is safer and clearer (to me):
```suggestion
                if (_charCount > 0)
```